tripal/tripal
# Restructure Prepare + add separate Import Type Collection

# Tripal 4 Core Dev Task

### Issue #1678 


### Tripal Version: 4

## Description


This PR makes the following changes:
1. Removes creating content types and attaching fields from the old prepare. This allows the prepare functionality to be focused entirely on preparing a specific chado instance. It can be run repeatedly for as many chado instances as you want.
2. Create a new "Import Type Collection" form reached from the Tripal > Page Structure menu. This allows admin to choose specific collections of content types to be created. Each collection listed is a single yaml file with the `tripal.tripalentitytype_collection.*.yml` file and they are auto-detected.
3. Update the existing *collection services to allow install of a single collection at a time. Previously the `install()` method installed all files available, but now it takes an array of collection ids and only installs those specified.
4. Break up the original massive `tripal.tripalentitytype_collection.default_chado.yml` and `tripal.tripalfield_collection.default_chado.yml` files into one file per category of data type. This is in preparation for all data types except "General" moving into their own extension modules.
5. Create a drush command allowing people to automatically import type collections from the command-line.
6. Add a call to the above drush command to tripaldocker to import the general type collection by default.

## Testing?




### Automated Testing

Upgraded existing Function TripalEntityTypeCollectionTest to a number of Kernel Tests which result in a total coverage of `tripal/src/Services/TripalEntityTypeCollection.php`: 94.62% (123 / 130) lines and 100% coverage for 4 / 6 methods.

![Screenshot 2023-11-20 at 4 09 13 PM](https://github.com/tripal/tripal/assets/1566301/4a95a578-92d1-4f50-8d4f-912d59d34516)

Existing Functional TripalFieldCollectionTest which had a total coverage of `tripal/src/Services/TripalFieldCollection.php`: 80.40% (160 / 199) lines and 100% coverage of 1/6 methods.

![Screenshot 2023-11-20 at 4 12 12 PM](https://github.com/tripal/tripal/assets/1566301/a000a7cf-a90a-4e40-be4b-99c877d263e9)

### Manual Testing

1. Create a fresh docker image + container on this branch.
2. Go to Admin > Tripal > Page Structure and confirm the General Types are the only ones present. NOTE: Until I finish the 2 TODOs in the description you will actually see no content types listed by default.
![Page Structure showing only general content types](https://github.com/tripal/tripal/assets/1566301/656778dc-3fec-44c7-bfec-ead2f14487e0)
3. Click on the new "Import type collection" button to see the new form. Confirm that all the expected types are listed as shown below.
![New Import type collection form](https://github.com/tripal/tripal/assets/1566301/bf6c6680-9575-480a-a68a-167cd28dd1cd)
4. Choose 1+ of these types and click the Import button. You will see a message saying that the Tripal job was submitted.
5. Run the tripal job (i.e. docker exec CONTAINER drush trp-run-jobs --username=drupaladmin`) and confirm that it completes without error. It will list each content type being created and also list each field being added to that content type. For example, you will see the following if you choose to import Genomic Content types.
![Terminal capture of import job](https://github.com/tripal/tripal/assets/1566301/4d9493ea-33eb-4641-8fbe-ed4aa7a96140)
6. Go back to the Page Structure page and confirm that the content types for that collection are now present. Click on manage fields for a few and confirm the fields are present. Add content for a few and confirm it adds without error.
![Page structure showing general and genomic content types](https://github.com/tripal/tripal/assets/1566301/51ba7091-5ca3-422b-9f92-c26b252b0e8b)
7. Click on the "Import type collection" button and select the same collection you already imported. Choose to import it and run the tripal job. Now you should see "Skip" in front of each one and still no errors.
8. Go to the command line and test the drush command. For example, `docker exec CONTAINER drush tripal:trp-import-types --username=drupaladmin --collection_id=germplasm_chado`
